### DISCUSSION TITLE
Checkbox questions with "Other" should allow adding multiple others

### DISCUSSION THREAD

**Discussion by LeaVerou** [Devographics/surveys#185] [General Survey Design]
There is a multiple choice question template that includes multiple tools with checkboxes, plus an "Other…" option, which shows a text field. Here is a screenshot of both states, from State of CSS 2023:

| Unselected | Selected |
|---|----|
| <img width="816" alt="image" src="https://github.com/Devographics/surveys/assets/175836/f901461f-d8df-482a-a72a-64b5d0acc7ec"> | <img width="816" alt="image" src="https://github.com/Devographics/surveys/assets/175836/7a487f62-9648-4736-9144-58f7a609b642"> |

The problem is that this only allows adding one "Other…" option, which forces the survey designer to pollute the survey with a ton of options (which biases the results) to minimize the need for users to need the "Other…" option.
Respondents can always enter comma-separated values, but this is both non-obvious, and becomes increasingly tedious as the number of values increases.

I present some UI ideas below to facilitate adding multiple custom options.

## General considerations

These apply across all UI ideas:
- Keyboard shortcuts to minimize homing between keyboard and pointing device:
	- Pressing Enter within a custom option should insert another custom option below it and focus it
	- Pressing Backspace within an empty custom option should delete it and focus the previous one
- The checkbox of a custom option is a distraction: when it is inconsistent with the presence or absence of text in the text field (i.e. checking it without entering text or not checking it while entering text), that tends to be a mistake. I suggest disabling it and using it entirely as output (checked if text is entered, unchecked if not). This would also improve keyboard UX by eliminating one tab stop.
- "Custom" may be better wording than "Other". It generalizes to [freeform lists](https://github.com/Devographics/surveys/discussions/194) better, and may imply less that the custom options are outliers.

## UI concept 1: "Add other" button

A potential UI for allowing multiple custom answers, without making the common case of entering one more tedious would be to convert the "Other…" option into an "Add Other…" button which inserts (and focuses) a text field option every time it’s clicked. 

| Before click | After click |
|---|----|
| <img width="811" alt="image" src="https://github.com/Devographics/surveys/assets/175836/4257756f-a3c8-40c5-be49-472dcc30d1a9"> | <img width="812" alt="image" src="https://github.com/Devographics/surveys/assets/175836/3f8750d7-2ee6-4aae-9d70-1832e43e34b2"> |

Notes:
- Empty fields are stripped from the data upon submission.
- When clicking "Add Other…", the new option inserted should also be selected and the text box should be focused

Pros:
- UX very similar to current one

Cons:
- While keyboard navigation is similar (new: tab to button, space to press it, type; old: tab to "Other" checkbox, space to check, tab to edit box), invoking buttons and going backwards feels less smooth
- Harder to generalize to [freeform lists](https://github.com/Devographics/surveys/discussions/194)
- UI still makes custom options feel like outliers, increasing echo chamber effects

## UI concept 2: Custom by default, auto-insert another

Update: I’ve now made a [high fidelity prototype for this](https://github.com/Devographics/surveys/discussions/185#discussioncomment-6843547)

https://github.com/Devographics/surveys/assets/175836/a8691fa0-5b1a-4cc9-b939-55ea51db8c8e

In this design, a custom field is present by default (potentially with styling to make it less prominent than it currently is). Typing in it inserts another (placeholder) one underneath, so entering more items is as simple as hitting tab (or enter). Hitting tab on an empty field does *not* insert another item.

Pros: 
- Invites more responses
- Better suited for freeform lists
- Very smooth, efficient keyboard navigation (nothing to check or press, just tabbing through like you do with regular choices)
- Less UI clutter (no need for buttons to add an option)

Cons:
- More prominent UI may not be desirable in cases where "Other" truly *is* an outlier
- It may be jarring to enter another placeholder item when you start typing? User testing can tell us.

**Comment by SachaG** [comment]
To clarify, the current "Other…" field accepts a freeform text string, which can contain one or multiple items. Maybe we can clarify this somehow in the current UI.

**Reply by ShaineRosewel** [reply]
I'm thinking if we can possibly add an instruction (at the same time force a rule on the input) for the respondents to separate multiple responses with a comma. If this is not too taxing on the respondent's end, it will help us also during the analysis (free form response for this item will be easier to clean).

**Comment by SachaG** [comment]
<img width="694" alt="Screenshot 2023-08-15 at 9 58 48" src="https://github.com/Devographics/surveys/assets/358832/d2df0b07-d802-4dfb-93c5-7adbcd7065f9">

This should work.

**Reply by LeaVerou** [reply]
This is better than nothing, but still discourages multiple responses, and makes it hard to edit more than 2-3. 😕
Are your concerns with the proposal in the OP primarily around engineering effort or user experience?

If eng effort: We could join the responses behind the scenes and still store as a single string, while exposing a UI that makes it easier to insert multiple items to end-users. Then we could use any separator we want when storing the joined string, e.g. something that is even less likely to appear in organic user input.

If user experience: We could keep the exact same UI for one "Other" response, and insert a "dormant" Other 2 item under it, that is only stored if edited. Similarly, if Other 2 is edited, a dormant Other 3: is added and so on. We should probably take checkboxes out of the tabbing order, and have the checked status of the checkbox be controlled entirely by whether text has been entered in the corresponding text field (making the checkbox primarily a feedback mechanism).

**Comment by LeaVerou** [comment]
I’ve fleshed out the interaction a bit more and made a high fidelity front-end prototype here: https://lea.verou.me/stateof/mocks/custom-options/

https://github.com/Devographics/surveys/assets/175836/a8691fa0-5b1a-4cc9-b939-55ea51db8c8e

What may not be super clear from the video:
- We typed two custom options, then deleted the second one using backspace to remove the item
- Then we tabbed away from the question, which removed all empty fields
- Then we typed two pain points, and tabbed away from that question, which removed the empty field at the end.

While the prototype supports both single-choice and multiple-choice questions, only the latter are discussed below, since the UX for single choice questions is very similar to the current one (does not support multiple custom options, since there is no reason to).

## Functionality

Both multiple and single choice questions can opt-in to allowing custom option(s). Single choice questions can only have one custom option, multiple choice questions have multiple by default. By default, any number of custom options is allowed, but this can be limited to a specific number. In that case, the number is communicated via the placeholder/aria-label. By default only one field is shown, but questions can opt-in to showing more, to encourage more question options.
Custom options have radios/checkboxes just like the predefined options, but they are disabled and get checked/unchecked automatically based on whether text has been entered.

It is possible to have questions without *any* predefined options, only custom ones. This gives us [freeform lists](https://github.com/Devographics/surveys/discussions/194) for free.

If there are no predefined options, the UI adapts in a few some small ways to best serve these use cases:
- Numbers (1, 2, 3...) are shown instead of checkboxes. This is currently done in the [template](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/index.html#L83), but could also be done with CSS alone.
- A `<textarea>` is used instead of an `<input>` We may later want to decouple this and use a separate `longform` setting.
- The placeholder text / aria label is different ("Item #1" instead of "Custom answer"). Perhaps there is some kind of placeholder text that would work for both ("freeform answer"?). 

## UX 

- Editing the last option inserts another one after it (but does not focus it)
- [Focusing away from a question](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/index.html#L26) removes all empty custom options except the first
- Keyboard navigation:
	- Backspace on an empty field deletes that option
	- [`<input>` only] Enter inserts an empty option under the current one and focuses it, unless an empty one already exists there in which case it just focuses it 
	- [`<input>` only] Up and Down arrow keys cycle through the options

## Architecture & Implementation

### Changes to question spec

Three extra fields:
- `allowCustom` (boolean): Enable custom answers, defaults to false.
- `initialCustomCount` (number): Number of custom answer fields shown initially, defaults to 1.
- `maxCustomCount` (number): Maximum number of custom answer fields, defaults to Infinity.

To minimize options, these could be combined into one `custom` field that could take either a boolean value to enable custom options with the default settings, or an object to customize the behavior (`{initial, max}`).

### Data model

Currently, each custom field has the same name, which should not change the current behavior when there is only one, backend code just needs to check whether there may also be multiple items and store an array in that case. It also means the logic works out of the box for single choice questions, with no conditional logic.

### Prototype code

- [Prototype files](https://github.com/LeaVerou/lea.verou.me/tree/main/files/stateof/mocks/custom-options)
- [Relevant part of the template](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/index.html#L68-L122)
- [Logic](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/models.js)
- [Keyboard interaction](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/index.js#L13)
- [CSS](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/style.postcss) (practically nonexistent)
- [Spec for sample questions used in prototype](https://github.com/LeaVerou/lea.verou.me/blob/main/files/stateof/mocks/custom-options/sample-questions.js)

## MVP

Compromises we can make to reduce engineering effort, in order (compromises to make earlier are listed first):
- Replacing checkboxes with counters when there are no predefined options (we can do it in CSS)
- Max number of custom options
- Keyboard navigation (Enter, Backspace, Up/Down arrows)
- Using a `<textarea>` and different placeholder/label when there are no predefined options
- Initial number of custom options

## Future

### Better support for user cases where custom options are outliers

Not all questions need custom answers at the same frequency. For some, the custom answers are the primary means of response, and the predefined answers exist just to get the respondent in the right headspace. In some, custom answers are needed only some of the time. And lastly, there are questions, that cannot be purely multiple choice, but rarely need a custom response.

Just like we can use an initial custom count > 1 to encourage custom answers *more*, we could also support an initial custom count of 0 to encourage them *less*. What would that look like? It could show a discreet "Custom…" button instead of a placeholder option. Alternatively, we could style the existing text field to just be less prominent in these questions.

### Sentiment chips on custom options

Ideally, custom options should also support [sentiment chips](https://github.com/Devographics/surveys/discussions/183) but that's a lot of work (both implementation and UX) so let's defer it for this survey.

**Reply by SachaG** [reply]
This is awesome and very thorough work as usual, and I will take the time to review it properly for future surveys, thanks!

I do have to confess that just on a personal level I would feel much more at ease if we could finish the outline first before we continue these kind of UI-focused explorations. As I've related in our other conversations, I have neither the time or the desire to focus on UI changes right now, so it'd be good to keep them at a minimum :)

**Reply by LeaVerou** [reply]
Hi @SachaG, I *am* keeping them at a minumum, this is one of the two high priority UI changes, as **they *directly* affect what the outline looks like**. Without it, we'd have to make some pretty bad compromises, and certain things would not be possible at all:

1. Questions such as [this](https://coda.io/d/_dTGBFYq175J/Missing-features-and-pain-points_suOQH#_luzbz) would not be possible, as the whole philosophy of this question is to prime participants with examples, but ultimately guide them towards entering their own answers. The current UI for entering custom answers is designed around custom answers being outliers, not the expected or a common response method.

2. Pain point questions (of which we currently have 9!) and these questions:
- [Which existing HTML features are you unable to use?](https://coda.io/d/_dTGBFYq175J/Missing-features-and-pain-points_suOQH#_lutn1)
- [Which components do you find yourself using across multiple projects?](https://coda.io/d/_dTGBFYq175J/Missing-features-and-pain-points_suOQH#_lupoC)

Would need to become freeform textareas, significantly reducing response rate (I’ve provided [citations for that here](https://github.com/Devographics/surveys/discussions/194#user-content-fn-1-09d48cbec1d629dfd35d3da406ccd8cf)), and making analysis harder.

3. We'd need to include a lot more options on several of the tooling questions, increasing survey fatigue. Case in point: the [Resources](https://survey.devographics.com/en-US/survey/state-of-css/2023/outline/12) page in State of CSS, in which some questions had to include so many options they needed …pagination:

<img width="821" alt="image" src="https://github.com/Devographics/surveys/assets/175836/2fcc7ea5-7c2c-428c-b138-ea0b2e32c20d">

With this we only need to include a few representative, well known examples, not the whole long tail of less popular stuff.

Please note that 1 and 2 are exactly the questions that give browser teams valuable data. It's **top priority** that we get these right, and we need to have prototypes in place for both this and the sentiment UI before user testing starts next week. Given that the prototypes I implemented are functional, all that is needed is to port them to the survey’s React-based infrastructure and hook up their options to the survey’s YML question specs. Hooking them up with the database doesn't need to happen before the user testing, but the UI needs to be there.

**Reply by SachaG** [reply]
I took the time to take another more thorough look at the prototype and read through this whole thread again to refresh my memory. 

Overall I think your prototype is great, and we might very well implement it (or something very close) at some point down the road. And while I totally agree that the current UIs for entering freeform data (namely the "Other…" checkbox, and plain freeform textfields) have their downsides, I think they are still functional enough for now. 

I might of course be wrong, but I don't recall these fields coming up as problems either in the feedback received through the "how can we improve this survey" questions on past surveys, or in feedback received via social media or GitHub issues. 

Given all that plus the fact that we have many other tasks to focus on (especially with this being the first time we're doing the State of HTML survey), I'm personally OK with postponing these UI improvements to next year's round of surveys.

**Reply by LeaVerou** [reply]
Hi @SachaG, I will re-iterate that this is not an enhancement, it is absolutely needed to implement the current outline. Please refer to my post above which went into detail about this.

Given that the prototype is fully functional, I would love to hear more about where the bottleneck is in terms of integrating it with the survey (even without actually storing data, or storing data as a comma-separated string or JSON serialized string or whatever), so we can work towards tweaks that may make it easier to implement.

> I might of course be wrong, but I don't recall these fields coming up as problems either in the feedback received through the "how can we improve this survey" questions on past surveys, or in feedback received via social media or GitHub issues.

First, users do not necessarily complain about usability issues, there are comparatively very few usability issues that users will be able to actually pinpoint and complain about — that's why we do user studies where we observe users (and even there certain issues do not show up). Second, this may have been an enhancement for previous surveys, but it is critical for the current outline, and affects some of the questions that are very important to the stakeholders who are actually funding this.

**Reply by SachaG** [reply]
I just don't understand what makes this such a high-priority item? The entire project for this survey was conceived with the previous question format, and that was never brought up as a potential issue until now. 

I also don't understand why this would have an impact on the survey outline. To be clear, we are talking about:

- A single "Other…" field vs multiple "Other Answer" fields
- A single freeform textfield vs multiple freeform textfields

Am I missing something? It seems to me like we can have the same survey with either UIs?

**Reply by LeaVerou** [reply]
Hi @SachaG,

> The entire project for this survey was conceived with the previous question format

My understanding is that neither @atopal nor I conceived the project in such a way that UI changes were not an option. 
Please note that State of HTML is different from State of JS or CSS, as it has a much broader scope, exactly because it actually extends beyond pure HTML to all sorts of web platform technologies (JS APIs, some CSS syntax, manifests, HTTP headers, a11y questions, etc etc). There are so many different stakeholders (several different browser teams, the developers that respond and later read the results, a variety of different standards groups and CGs etc), each with different goals and interests. The amount of content needed for all of them to get at least *some* useful data is simply a lot higher than in the other surveys. If we are only able to use existing UI elements, it would simply not be possible to include what is needed and maintain a reasonable length.

> that was never brought up as a potential issue until now.

I identified this as top priority (second only to sentiment) on August 17th, 2 weeks ago.

> I just don't understand what makes this such a high-priority item? 
> I also don't understand why this would have an impact on the survey outline. [...]

I explained [above](https://github.com/Devographics/surveys/discussions/185#discussioncomment-6852289) why we need this, but let me rephrase, in case my earlier comment buried the lede:
1. It's not just about one vs multiple answers. **It's about how much the UI encourages custom answers.** The current UI is designed around custom answers being outliers, and requires the questions to be designed accordingly. For some questions (e.g. tooling), it is possible to just include a ton of predefined answers so that "Other" is rarely if ever needed. I can see the argument that for *those* questions, this may indeed just be an enhancement. Though even for those, a) having super long questions increases survey fatigue, which is exactly what we *desperately* need to reduce here and b) getting to the point where a question needs pagination biases the results. However, for other questions, where we want to encourage a large % of answers to be custom (e.g. missing elements question), having a single "Other" field that requires an extra click and is not on the same "level" as other answers does not give us the kind of interaction that is necessary to collect the data we need.
2. For questions without predefined options (freeform lists), I have provided [citations to studies](https://github.com/Devographics/surveys/discussions/194#user-content-fn-1-44e7e195977066dd591a3899d70873e3) showing how freeform lists encourage **significantly** more responses. Given that these are exactly the questions that are *critical* for browser teams, why would we not do something as easy as this to encourage responses (plus make analysis a little easier)?
3. If this were a ton of work to implement to the point it could actually delay launch (as @atopal is worried below), I would understand maybe deferring it. But at this point, there is a fully functional prototype, and all it needs is to be ported to React. It doesn't even necessarily require data model changes (if we serialize the answers as a single string and store that). In fact, I feel the time we're spending debating whether to attempt this or not in total might well be longer than what porting it would take. 😃 If that is not accurate, could you please elaborate on what is tricky?

**Reply by SachaG** [reply]
> having a single "Other" field that requires an extra click and is not on the same "level" as other answers does not give us the kind of interaction that is necessary to collect the data we need.

Based on what grounds?

I feel like you're putting the burden of proof on me to prove that changing the UI is unnecessary, but that seems backwards to me. The burden of proof should be on you to support your claim that the current UI is inadequate.

Again I am not saying your proposals are not improvements, I am saying that unless we have evidence that the current UI will not let us fulfill the survey's goals, then as far as I'm concerned the bar for justifying these UI changes *at this time* is not cleared. 

On the other hand, what we **do** have evidence of, is that the current UI has worked for numerous surveys in the past, and that the other stakeholders in this project have not raised any issues about it so far. 

> In fact, I feel the time we're spending debating whether to attempt this or not in total might well be longer than what porting it would take.

I agree that time spent debating this is not well spent, and a better use of time would be working on the outline – which unlike this is core to the survey.

**Reply by LeaVerou** [reply]
My point was **not** that the current UI is inadequate, just that it's designed with different use cases in mind: those where custom answers are outliers. Designing questions so that custom answers are outliers is something that definitely affects the outline, and is not always even possible.

**Reply by ShaineRosewel** [reply]
Adding multiple, separate fields for other inputs will definitely help us during the analysis phase. There are users who input multiple ideas, say, when asked about pain points. When analyzing, this means we'd have to separate different ideas if we only have a single input field. It will be a big help if these multiple ideas are already separated to begin with. Hence, I support multiple input fields for others and free form responses.

**Reply by ShaineRosewel** [reply]
In the past, I have seen lengthy answers with different items being tackled while others with a brief response in one input field. How do we extract insights from these? We might want to summarize lengthy answers through the use of existing models. However, summarizing long ones might mean we lose some information (since existing models sometimes get confused with lengthy statements with multiple ideas). We may not need to worry about losing information if multiple ideas are already separated in different input fields to begin with.

In terms of say, other magazines respondents read, if they want to input three more items in the 'others', telling them to separate entries with commas is an option but we are giving them an additional instruction to follow; also, we cannot force them to follow (eg, I used semicolon to enter multiple response and still it got recorded). This is still possible if we have multiple input fields for multiple answers. But, the idea is we took away the effort from the respondents to follow an additional instruction.

**Comment by atopal** [comment]
It looks like we're all agreed that the changes would be beneficial, and it comes down to how much benefit and how much cost. Sacha, can you comment on what implementing this would mean for the State of HTML survey? Would we risk delaying it, or would this mean a significant amount of work on your end? Lea, do we have indicators that people are more likely to add significantly more information with this UI change?
